### PR TITLE
[HACKATHON] Email notification option on gem2s loading screen

### PR DESCRIPTION
**Pull request by StefanBabukov**
# Background
#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [ ] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [ ] Deployed to staging
- [ ] Passed integration tests 

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/api/route-services/gem2s.js [hunk 1]
```diff
@@ -24,7 +30,7 @@ pipelineHook.register('uploadToAWS', [saveProcessingConfigFromGem2s, runQCPipeli
 class Gem2sService {
   static async sendUpdateToSubscribed(experimentId, message, io) {
     const statusRes = await getPipelineStatus(experimentId, constants.GEM2S_PROCESS_NAME);
-
+    const { status } = statusRes.gem2s;
```
Comment (kafkasl) [inline src/api/route-services/gem2s.js:33]: this should be closer to where it's used

File: src/api/route-services/__mocks__/experiment.js [hunk 2]
```diff
@@ -4,6 +4,14 @@ const mockExperimentData = jest.fn((experimentId) => new Promise((resolve) => {
   resolve({
     experimentId,
     experimentName: 'my mocky name',
+    meta: {
+      pipeline: {
+        stateMachineArn: 'qwerty',
```
Comment (aerlaut) [inline src/api/route-services/__mocks__/experiment.js]: can you switch this to a more realistic ARN example?

File: src/api/route-services/__mocks__/experiment.js [hunk 3]
```diff
@@ -4,6 +4,14 @@ const mockExperimentData = jest.fn((experimentId) => new Promise((resolve) => {
...
+      },
+      gem2s: {
+        stateMachineArn: 'asdfgh',
```
Comment (aerlaut) [inline src/api/route-services/__mocks__/experiment.js]: This too

File: src/api/route-services/experimentHelpers.js [hunk 4]
```diff
@@ -71,12 +71,13 @@ const getShallowAttrsUpdateParams = (body) => {
...
     description: body.description,
     sampleIds: body.sampleIds,
+    notifyByEmail: body.notifyByEmail,
   };
 
   const objectToMarshall = {};
   const updateExpressionList = Object.entries(dataToUpdate).reduce(
     (acc, [key, val]) => {
-      if (!val) {
+      if (val === undefined) {
```
Comment (aerlaut) [inline src/api/route-services/experimentHelpers.js]: why the change to `=== undefined`? are there cases where `val` is `null` or `0` where we want to still insert  `${key} = ${expressionKey}`?

File: src/utils/hooks/run-qc-pipeline.js [hunk 5]
```diff
@@ -5,8 +5,9 @@ const { createQCPipeline } = require('../../api/general-services/pipeline-manage
 const experimentService = new ExperimentService();
 
 const runQCPipeline = async (payload) => {
-  const { experimentId, authJWT } = payload;
-
+  const { experimentId } = payload;
+  // we need to change this once we rework the pipeline message response
+  const authJWT = payload.authJWT || payload.input.authJWT;
```
Comment (aerlaut) [inline src/utils/hooks/run-qc-pipeline.js:10]: What is the issue? some pipelines have `authJWT` under `input` while others don't?

File: src/utils/hooks/send-notification.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+const getPipelineStatus = require('../../api/general-services/pipeline-status');
+const { SUCCEEDED, FAILED, QC_PROCESS_NAME } = require('../../api/general-services/pipeline-manage/constants');
+const sendEmail = require('../send-email');
+const sendFailedSlackMessage = require('../send-failed-slack-message');
+const ExperimentService = require('../../api/route-services/experiment');
+const config = require('../../config');
+
+const sendNotification = async (message) => {
+  const { authJWT, processName: process } = message.input;
+  if (authJWT) {
```
Comment (aerlaut) [inline src/utils/hooks/send-notification.js]: instead of enclosing all the block in `if(authJWT)`, would it be clearer if you handle the case for `if(!authJWT)` here?

File: src/api/route-services/experimentHelpers.js (hunk 4 above)
Comment (StefanBabukov) [reply src/api/route-services/experimentHelpers.js]: when i set notifyByEmail to false this ignores it

File: src/utils/hooks/send-notification.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      }
+      if (experiment.notifyByEmail && (process === QC_PROCESS_NAME || status === FAILED)) {
+        try {
+          sendEmail(message.experimentId, status, user);
+        } catch (e) {
+          console.error('Error sending email ', e);
+        }
+      }
+    }
+  }
```
Comment (aerlaut) [inline src/utils/hooks/send-notification.js]: I think you can make this nested if blocks clearer too

File: src/utils/hooks/run-qc-pipeline.js (hunk 5 above)
Comment (StefanBabukov) [reply src/utils/hooks/run-qc-pipeline.js:10]: Yeah we need to re-work the pipeline message that is returned. There are duplicate values in many places and sometimes the token is under input and sometimes its not.

File: src/utils/send-email.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  SUCCEEDED,
+} = require('../api/general-services/pipeline-manage/constants');
+// const ExperimentService = require('../api/route-services/experiment');
+
+const sendEmail = (experimentId, status, user) => {
+  const ses = new AWS.SES({
+    region: config.awsRegion,
+    apiVersion: '2010-12-01',
+  });
+  // const experiment = await (new ExperimentService()).getExperimentData(experimentId);
```
Comment (aerlaut) [inline src/utils/send-email.js]: don't forget to remove this if not needed

File: src/utils/send-email.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+      },
+      Subject: {
+        Charset: 'UTF-8',
+        Data: 'Biomage Pipeline status',
+      },
+    },
+    Source: 'notification@biomage.net',
+  };
+
+  return ses.sendEmail(params).promise().then((response) => console.log(response));
```
Comment (aerlaut) [inline src/utils/send-email.js]: can you make `sendEmail` into an async function? 
Do we need the `console.log(response)`? Or do we need only a confirmation log that the email has been sent?

File: src/utils/send-email.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        <head>
+        </head>
+        <body>
+            <h3>Hello ${firstname},</h3>
+            <p>Thanks for uploading a dataset to Cellscope! <br/>
+              ${status === SUCCEEDED ? successMessage : failMessage}<br/><br/>
+                The Biomage Team 
+            <p/>
+        </body>
+      </html>`;
```
Comment (aerlaut) [inline src/utils/send-email.js]: Can you encapsulate these into a function that outputs `params`? maybe `buildEmailBody(....)`?

File: src/utils/send-failed-slack-message.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
+const atob = require('atob');
+
+const sendFailedSlackMessage = async (message, user, experiment) => {
+  const { experimentId } = message;
+  // this needs to change once we change the name to qc in experiments/meta
```
Comment (aerlaut) [inline src/utils/send-failed-slack-message.js]: change this comment to BOLD or make something to make it more obvious

File: tests/api/route-services/gem2s.test.js [hunk 12]
```diff
@@ -132,32 +133,12 @@ describe('gem2s', () => {
...
-    const emitParamsChannel = mockedSocketsEmit.mock.calls[0][0];
-    expect(mockedSocketsEmit).toHaveBeenCalled();
-
-    // Emitted to the correct channel
-    expect(emitParamsChannel).toMatch(experimentId);
-    expect(mockedSocketsEmit).toMatchSnapshot();
-  });
-
-
-  it('gem2sResponse - Should return message if message is valid', async () => {
```
Comment (aerlaut) [inline tests/api/route-services/gem2s.test.js:148]: great catch

File: tests/api/route-services/gem2s.test.js [hunk 13]
```diff
@@ -201,6 +182,10 @@ describe('gem2s', () => {
       response: {
         error: errorText,
       },
+      input: {
+        authJWT: 'Bearer ayyylmaoxd',
```
Comment (aerlaut) [inline tests/api/route-services/gem2s.test.js]: make authJWT look more real?

File: tests/test-utils/constants.js [hunk 14]
```diff
@@ -52,8 +52,11 @@ const S3_WORKER_RESULT = {
     },
   },
 };
-
-
+const USER = {
+  name: 'John Cena',
+  email: 'johncena@gmail.com',
+  sub: 'useridayylmao',
```
Comment (aerlaut) [inline tests/test-utils/constants.js:58]: Why John Cena? not Triple H?

File: tests/utils/hookRunner.test.js [hunk 15]
```diff
@@ -5,18 +5,13 @@ describe('HookRunner', () => {
     const runner = new HookRunner();
 
     expect(runner.hooks).toEqual({});
+    expect(runner.results).toEqual({});
   });
 
-  it('should register single hooks properly', () => {
```
Comment (aerlaut) [inline tests/utils/hookRunner.test.js:10]: Why don't we support single hooks anymore?

File: tests/utils/hookRunner.test.js [hunk 16]
```diff
@@ -56,11 +50,53 @@ describe('HookRunner', () => {
...
 
     await runner.run({ taskName: 'event' });
 
     expect(runner.results.event).toEqual([fn1(), fn2()]);
   });
+
+  it('should run multiple hooks with a global one', async () => {
+    const fn1 = jest.fn(() => 1);
+    const fn2 = jest.fn(() => 2);
+    const fn3 = jest.fn(() => 3);
```
Comment (aerlaut) [inline tests/utils/hookRunner.test.js]: maybe change the name to be easier to understand? I got a little bit confused when seeing `runner.register('event2', [fn3])`

File: tests/utils/hooks/send-notification.test.js [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+  const message = {
+    experimentId: 'mockexp',
+    input: {
+      authJWT: 'somejwt',
+      processName: 'gem2s',
+      taskName: 'task1',
+    },
+  };
+
+  global.fetch = jest.fn(() => Promise.resolve({ ok: true }));
```
Comment (aerlaut) [inline tests/utils/hooks/send-notification.test.js]: don't we have `jest-fetch-mock`?

File: tests/utils/hooks/send-notification.test.js [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+      gem2s: {
+        status: FAILED,
+      },
+    });
+    await sendNotification(message);
+    expect(sendFailedSlackMessage).toHaveBeenCalledTimes(1);
+    expect(sendEmail).toHaveBeenCalledTimes(0);
+  });
+
+  it('Sends email if user toggled notifications on failed process', async () => {
```
Comment (aerlaut) [inline tests/utils/hooks/send-notification.test.js]: maybe `Send email and Slack message....`

File: src/utils/send-email.js (hunk 10 above)
Comment (aerlaut) [reply src/utils/send-email.js]: it would make it easier too if we want to extend this functionality to send other types of emails.

File: tests/test-utils/constants.js (hunk 14 above)
Comment (StefanBabukov) [reply tests/test-utils/constants.js:58]: hahaha

File: tests/utils/hookRunner.test.js (hunk 15 above)
Comment (StefanBabukov) [reply tests/utils/hookRunner.test.js:10]: Pol moved them into an array so i guess this test is redundant?

File: tests/utils/hooks/send-notification.test.js (hunk 17 above)
Comment (StefanBabukov) [reply tests/utils/hooks/send-notification.test.js]: This was an easier 1 line solution without imports, I don't need complex fetch logic. What do you think?

Comment (aerlaut) [reply tests/utils/hooks/send-notification.test.js]: Under the hood, this is actually the way `jest-fetch-mock` is implemented. If we have the library and it is used in the other tests as well, I think it's better to use `jest-fetch-mock`.

File: src/utils/send-email.js [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+  SUCCEEDED,
+} = require('../api/general-services/pipeline-manage/constants');
+
+const sendEmail = (experimentId, status, user) => {
+  const ses = new AWS.SES({
+    region: config.awsRegion,
+    apiVersion: '2010-12-01',
+  });
+
+  const buildEmailBody = () => {
```
Comment (aerlaut) [inline src/utils/send-email.js]: Can you refactor out this component (put it out of this `sendEmail` function) and rename it to something that is more specific? (e.g. `buildPipelineStatusEmailBody`). The function should also not rely on global `user`, `status` and `experimentId` and take them in as parameters instead,

File: src/utils/hooks/send-notification.js [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    return;
+  }
+  const user = await authenticationMiddlewareSocketIO(authJWT);
+
+  const { experimentId } = message;
+  const statusRes = await getPipelineStatus(experimentId, process);
+  const experiment = await (new ExperimentService()).getExperimentData(experimentId);
+  const { status } = statusRes[process];
+
+  if (status === FAILED && ['production', 'staging', 'test'].includes(config.clusterEnv)) {
```
Comment (kafkasl) [inline src/utils/hooks/send-notification.js]: remove staging before merging

**Review by kafkasl — APPROVED** [review]
